### PR TITLE
feat(nebula_ros): all sensors json schema support

### PR DESCRIPTION
**Pull request by amadeuszsz**
## PR Type

- New Feature

## Related Links

* Issue: https://github.com/tier4/nebula/issues/63
* Schema discussion: https://github.com/orgs/autowarefoundation/discussions/3371
* Schema docs: https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/parameters/
* CI PR: https://github.com/autowarefoundation/autoware-github-actions/pull/294

## Description

The PR adds json schema to nebula which is current way of handling param files in Autoware environment.

Old proposal aka inheritance. Due to our abstraction, we can't use `additionalProperties=false`. Sub schemas are pushed to the same parent key `ros__parameters`, therefore they would block itself in case of not allowing extra parameters. With `additionalProperties=true`, users can provide extra parameters to `.param.yaml` but at the end they not affect runtime.
schema

New proposal aka namespaces. `additionalProperties=false` possible, user can't provide extra parameters. In addition, less maintenance - parameters are placed in namespaces and new sensors will not brake architecture. The final `.json.schema` points specific parameters from sub schema files, it's possible to overwrite fields (e.g. if we need to use  `rotation_speed` but it's required to define maximum).
schema_v2

In general, schema considers only two numeric types: `number` and `integer`. For floating point number and type of `number`, the user can provide `integer` value as well. This will throw exception in node constructor. From schema level, we can't force user to use floating point number. We could provide extra method for declaring floating point number instead of `declare_parameter`. The idea would be to load parameter with dynamic typing, check the type, cast it to `double` and set it again.

## Review Procedure

1. Install json-schema
```bash
pip install check-jsonschema
```
2. Copy [CI script](https://github.com/amadeuszsz/autoware-github-actions/blob/feat/allow-nested-config/json-schema-check/validate_json_schemas.py) locally, e.g. to package root.
3. Test nebula param files.
```bash
cd <nebula-root>
python validate_json_schemas.py
```
4. Test VS Code extension.
* Install YAML extension (redhat.vscode-yaml).
* Update your `.vscode/settings.json`, e.g.:
```json
    "yaml.schemas": {
        "src/vendor/nebula/nebula_ros/schema/VLP16.schema.json": "lidar/velodyne/VLP16.param.yaml"
    }
```
* Check if works.

Old proposal.
schema-vscode

New proposal. All the parameters are generated automatically - no need to point abstractions.
schema-vscode_v2

## Remarks

* Since this PR is made on top of one development branch, please wait for https://github.com/tier4/nebula/pull/148 merged.
* The CI requires merge in external repository first. If PR will be rejected, we might add CI script into nebula package. 

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

File: nebula_ros/src/hesai/hesai_ros_wrapper.cpp [hunk 1]
```diff
@@ -103,30 +103,30 @@ nebula::Status HesaiRosWrapper::DeclareAndGetSensorConfigParams()
       default_value = 600;
     }
     config.rotation_speed =
-      declare_parameter<uint16_t>("rotation_speed", default_value, descriptor);
```
Comment (mojomex) [inline nebula_ros/src/hesai/hesai_ros_wrapper.cpp:106]: Please remove `deault_value` in lines `94`, `99` and `103` as well

File: nebula_ros/schema/Pandar128E4X.schema.json [hunk 2]
```diff
@@ -0,0 +1,132 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne Pandar128E4X parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/Pandar128E4X.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai Pandar128E4X parameters.",
```

File: nebula_ros/schema/Pandar40P.schema.json [hunk 3]
```diff
@@ -0,0 +1,128 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne Pandar40P parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/Pandar40P.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai Pandar40P parameters.",
```

File: nebula_ros/schema/Pandar64.schema.json [hunk 4]
```diff
@@ -0,0 +1,124 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne Pandar64 parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/Pandar64.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai Pandar64 parameters.",
```

File: nebula_ros/schema/PandarAT128.schema.json [hunk 5]
```diff
@@ -0,0 +1,144 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne PandarAT128 parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/PandarAT128.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai PandarAT128 parameters.",
```

File: nebula_ros/schema/PandarQT128.schema.json [hunk 6]
```diff
@@ -0,0 +1,132 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne PandarQT128 parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/PandarQT128.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai PandarQT128 parameters.",
```

File: nebula_ros/schema/PandarQT64.schema.json [hunk 7]
```diff
@@ -0,0 +1,128 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne PandarQT64 parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/PandarQT64.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai PandarQT64 parameters.",
```

File: nebula_ros/schema/PandarXT32.schema.json [hunk 8]
```diff
@@ -0,0 +1,132 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne PandarXT32 parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/PandarXT32.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai PandarXT32 parameters.",
```

File: nebula_ros/schema/PandarXT32M.schema.json [hunk 9]
```diff
@@ -0,0 +1,132 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "LiDAR Velodyne PandarXT32M parameters.",
```
Comment (mojomex) [inline nebula_ros/schema/PandarXT32M.schema.json:3]:
```suggestion
  "title": "LiDAR Hesai PandarXT32M parameters.",
```

File: nebula_ros/schema/example.schema.template.json [hunk 10]
```diff
@@ -0,0 +1,62 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "<sensor_type> <sensor_vendor> <sensor_model> parameters.",
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply nebula_ros/schema/example.schema.template.json:3]: Agree, but this follows config directory order

File: nebula_ros/src/hesai/hesai_ros_wrapper.cpp (hunk 1 above)
Comment (amadeuszsz) [reply nebula_ros/src/hesai/hesai_ros_wrapper.cpp:106]: https://github.com/tier4/nebula/pull/155/commits/16a2a493d350c00c877d60e7eb4e4cfbde1cd13b

File: nebula_ros/config/lidar/hesai/PandarAT128.param.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    packet_mtu_size: 1500
+    launch_hw: true
+    setup_sensor: true
+    frame_id: hesai
+    diag_span: 1000
+    correction_file: $(find-pkg-share nebula_decoders)/calibration/hesai/$(var sensor_model).dat
+    min_range: 0.3
+    max_range: 300.0
+    cloud_min_angle: 0
+    cloud_max_angle: 360
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/PandarAT128.param.yaml:16]:
```suggestion
    cloud_max_angle: 150
```

File: nebula_ros/config/lidar/hesai/PandarAT128.param.yaml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    gnss_port: 10110
+    packet_mtu_size: 1500
+    launch_hw: true
+    setup_sensor: true
+    frame_id: hesai
+    diag_span: 1000
+    correction_file: $(find-pkg-share nebula_decoders)/calibration/hesai/$(var sensor_model).dat
+    min_range: 0.3
+    max_range: 300.0
+    cloud_min_angle: 0
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/PandarAT128.param.yaml:15]:
```suggestion
    cloud_min_angle: 30
```

File: nebula_ros/config/radar/continental/ARS548.param.yaml [hunk 13]
```diff
@@ -1,13 +1,14 @@
 /**:
   ros__parameters:
-    sensor_model: "ARS548"
-    frame_id: "continental"
-    base_frame: "base_link"
-    object_frame: "base_link"
-    sensor_ip: "10.13.1.114"
     host_ip: "10.13.1.166"
+    sensor_ip: "10.13.1.114"
+    data_port: 2368
```
Comment (mojomex) [inline nebula_ros/config/radar/continental/ARS548.param.yaml:5]:
```suggestion
    data_port: 42102
```

File: nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    max_range: 300.0
+    cloud_min_angle: 0 
+    cloud_max_angle: 360
+    scan_phase: 0.0
+    sensor_model: Pandar128E4X
+    calibration_file: $(find-pkg-share nebula_decoders)/calibration/hesai/$(var sensor_model).csv
+    rotation_speed: 600
+    return_mode: Dual
+    ptp_profile: "1588v2"
+    ptp_domain: 0
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml:22]: OT128 (128E4X) only supports automotive profile, set it as default
```suggestion
    ptp_profile: automotive
```

File: nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    ptp_transport_type: UDP
+    ptp_switch_type: TSN
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml:24]: Automotive profile requires L2 transport, set it as default
```suggestion
    ptp_transport_type: L2
```

File: nebula_ros/schema/sub/hardware.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    "setup_sensor": {
+      "type": "boolean",
+      "default": true,
+      "readOnly": true,
+      "description": "Enable sensor setup on hardware driver."
+    },
+    "retry_hw": {
+      "type": "boolean",
+      "default": true,
+      "description": "hw driver startup retry (false when using pcap)."
```
Comment (mojomex) [inline nebula_ros/schema/sub/hardware.json:29]: Not inherently related to PCAPs, a more generic description like the one below would be preferable.
```suggestion
      "description": "Whether TCP connections are retried on failure or the driver should instead exit."
```

File: nebula_ros/schema/sub/hardware.json [hunk 17]
```diff
@@ -0,0 +1,32 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Hardware parameters.",
+  "type": "object",
+  "definitions": {
+    "launch_hw": {
+      "type": "boolean",
+      "readOnly": false,
+      "description": "Launch sensor hardware."
```
Comment (mojomex) [inline nebula_ros/schema/sub/hardware.json:9]: A more precise description would be preferable.
```suggestion
      "description": "Whether network sockets should be opened or not. If disabled, replay from NebulaPackets messages is enabled automatically."
```

File: nebula_ros/schema/sub/topic.json [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    "base_frame": {
+      "type": "string",
+      "default": "base_link",
+      "readOnly": true,
+      "description": "Frame with true sensor pose. The final sensor data is transformed by base_frame -> frame_id."
+    },
+    "diag_span": {
+      "type": "integer",
+      "default": 1000,
+      "minimum": 0,
```
Comment (mojomex) [inline nebula_ros/schema/sub/topic.json:21]: `0 ms` timer period would not be any good, let's set it to a `> 0` value (`1 ms` is technically valid but still insane....)
```suggestion
      "minimum": 1,
```

File: nebula_ros/schema/Pandar128E4X.schema.json [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+            "Strongest",
+            "LastStrongest",
+            "First",
+            "LastFirst",
+            "FirstStrongest",
+            "Dual"
+          ]
+        },
+        "ptp_profile": {
+          "$ref": "sub/communication.json#/definitions/ptp_profile"
```
Comment (mojomex) [inline nebula_ros/schema/Pandar128E4X.schema.json:73]: Only allow `automotive` profile for OT128
```suggestion
          "$ref": "sub/communication.json#/definitions/ptp_profile",
          "enum": [
            "automotive"
          ]
```

File: nebula_ros/schema/Pandar128E4X.schema.json [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+        },
+        "ptp_domain": {
+          "$ref": "sub/communication.json#/definitions/ptp_domain"
+        },
+        "ptp_transport_type": {
+          "$ref": "sub/communication.json#/definitions/ptp_transport_type"
```
Comment (mojomex) [inline nebula_ros/schema/Pandar128E4X.schema.json:79]: Only allow `L2` transport for OT128 (see datasheet).
```suggestion
          "$ref": "sub/communication.json#/definitions/ptp_transport_type",
          "enum": [
            "L2"
          ]
```

**Review by mojomex — CHANGES_REQUESTED** [review]
@amadeuszsz @vividf Looking really good! Thank you for your work!
I have added a few suggestions here and there, mainly related to description texts and OT128's supported PTP settings. Once these are implemented the PR should be good to merge.

File: nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    min_range: 0.3
+    max_range: 300.0
+    cloud_min_angle: 0 
+    cloud_max_angle: 360
+    scan_phase: 0.0
+    sensor_model: Pandar128E4X
+    calibration_file: $(find-pkg-share nebula_decoders)/calibration/hesai/$(var sensor_model).csv
+    rotation_speed: 600
+    return_mode: Dual
+    ptp_profile: "1588v2"
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml:21]: OT128 only supports `automotive` profile
```suggestion
    ptp_profile: automotive
```

File: nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    ptp_domain: 0
+    ptp_transport_type: UDP
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/Pandar128E4X.param.yaml:23]: OT128 only supports `L2` transport
```suggestion
    ptp_transport_type: L2
```

**Review by mojomex — CHANGES_REQUESTED** [review]
(some suggestions were incorrectly referencing the wrong line, deleted them above and added the corrected ones here)

**Review by mojomex — APPROVED** [review]
LGTM!

**Pull request merged**